fastly/cloudfiles.go

```diff
@@ -0,0 +1,255 @@
+	AccessKey         string     `mapstructure:"access_key"`
+	BucketName        string     `mapstructure:"bucket_name"`
+	Path              string     `mapstructure:"path"`
+	Region            string     `mapstructure:"region"`
+	Placement         string     `mapstructure:"placement"`
+	Period            uint       `mapstructure:"period"`
+	GzipLevel         uint       `mapstructure:"gzip_level"`
+	Format            string     `mapstructure:"format"`
+	FormatVersion     uint       `mapstructure:"format_version"`
+	ResponseCondition string     `mapstructure:"response_condition"`
+	MessageType       string     `mapstructure:"message_type"`
+	TimestampFormat   string     `mapstructure:"timestamp_format"`
+	PublicKey         string     `mapstructure:"public_key"`
+	CreatedAt         *time.Time `mapstructure:"created_at"`
+	UpdatedAt         *time.Time `mapstructure:"updated_at"`
+	DeletedAt         *time.Time `mapstructure:"deleted_at"`
+}
+
+// cloudfilesByName is a sortable list of Cloudfiles.
+type cloudfilesByName []*Cloudfiles
+
+// Len, Swap, and Less implement the sortable interface.
+func (c cloudfilesByName) Len() int      { return len(c) }
+func (c cloudfilesByName) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+func (c cloudfilesByName) Less(i, j int) bool {
+	return c[i].Name < c[j].Name
+}
+
+// ListCloudfilesInput is used as input to the ListCloudfiles function.
+type ListCloudfilesInput struct {
+	// Service is the ID of the service (required).
+	Service string
+
+	// Version is the specific configuration version (required).
+	Version int
+}
+
+// ListCloudfiles returns the list of Cloudfiles for the configuration version.
+func (c *Client) ListCloudfiles(i *ListCloudfilesInput) ([]*Cloudfiles, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/logging/cloudfiles", i.Service, i.Version)
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var cloudfiles []*Cloudfiles
+	if err := decodeJSON(&cloudfiles, resp.Body); err != nil {
+		return nil, err
+	}
+	sort.Stable(cloudfilesByName(cloudfiles))
+	return cloudfiles, nil
+}
+
+// CreateCloudfilesInput is used as input to the CreateCloudfiles function.
+type CreateCloudfilesInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	Name              string `form:"name,omitempty"`
+	User              string `form:"user,omitempty"`
+	AccessKey         string `form:"access_key,omitempty"`
+	BucketName        string `form:"bucket_name,omitempty"`
+	Path              string `form:"path,omitempty"`
+	Region            string `form:"region,omitempty"`
+	Placement         string `form:"placement,omitempty"`
+	Period            uint   `form:"period,omitempty"`
+	GzipLevel         uint   `form:"gzip_level,omitempty"`
```

Comment:
We should discuss how we want to handle scalar values and `omitempty` in the update input structs. We've recently started to move other update structs over to pointer values. The problem is the combination of Go's type zero values, JSON marshalling with the `omitempty` struct field tag, and a users intent to actually set a zero value. This blog post does a much better job at explaining the problem and well worth a read: https://willnorris.com/2014/05/go-rest-apis-and-pointers/ .

Therefore this current implementation would make it impossible to update these values to their zero values i.e. `0` in this case. A more in-depth explanation and discussion of recent decisions made in the library can be found in this PR: https://github.com/fastly/go-fastly/pull/142#discussion_r369001840